src/Tribe/Repositories/Traits/Post_Tickets.php

```diff
@@ -0,0 +1,292 @@
+			$request_codes[] = (array) $currency->get_symbol_codes( $request_symbol );
+		}
+		$request_codes = array_unique( call_user_func_array( 'array_merge', $request_codes ) );
+
+		if ( empty( $request_codes ) ) {
+			$reason = 'The specified currency symbol or ISO code is not supported.';
+			throw Tribe__Repository__Void_Query_Exception::because_the_query_would_yield_no_results( $reason );
+		}
+
+		// Compile a list of ticket providers that are active and use one of the requested ISO codes.
+		$request_providers = [];
+
+		if ( array_intersect( $request_codes, (array) $currency->get_currency_code() ) ) {
+			$request_providers[] = 'tpp';
+		}
+
+		$providers = [
+			'eddticket' => 'Tribe__Tickets_Plus__Commerce__EDD__Main',
+			'wooticket' => 'Tribe__Tickets_Plus__Commerce__WooCommerce__Main',
+		];
+
+		foreach ( $providers as $slug => $provider ) {
+			if ( ! class_exists( $provider ) ) {
+				continue;
+			}
+
+			$provider_symbol = $currency->get_provider_symbol( $provider );
+
+			if ( ! is_string( $provider_symbol ) ) {
+				continue;
+			}
+
+			$provider_codes = (array) $currency->get_symbol_codes( $provider_symbol );
+
+			if ( array_intersect( $request_codes, $provider_codes ) ) {
+				// This provider uses one of the request ISO codes, use it.
+				$request_providers[] = $slug;
+			}
+		}
+
+		if ( empty( $request_providers ) ) {
+			$reason = 'No ticket provider uses the specified currency symbol or ISO code.';
+			throw Tribe__Repository__Void_Query_Exception::because_the_query_would_yield_no_results( $reason );
+		}
+
+		$providers_regex = sprintf( '(%s)', implode( '|', $request_providers ) );
+
+		global $wpdb;
+		$prefix = 'by_cost_currency_symbol_';
+		// Join to the meta that relates tickets to events but only for the providers that have the required symbols.
+		$repo->join_clause( "JOIN {$wpdb->postmeta} {$prefix}_ticket_event 
+			ON ( 
+				{$prefix}_ticket_event.meta_value = {$wpdb->posts}.ID
+				AND {$prefix}_ticket_event.meta_key REGEXP '^_tribe_{$providers_regex}_for_event\$' 
+			)" );
+	}
+
+	/**
+	 * Filters events to include only those that match the provided ticket state.
+	 *
+	 * This does NOT include RSVPs or events that have a cost assigned via the
+	 * cost custom field.
+	 *
+	 * @since TBD
+	 *
+	 * @param bool $has_tickets Indicates if the event should have ticket types attached to it or not.
+	 */
+	public function filter_by_has_tickets( $has_tickets = true ) {
+		$repo = $this;
+
+		// If the repo is decorated, use that.
+		if ( ! empty( $repo ) ) {
+			$repo = $this->decorated;
+		}
+
+		global $wpdb;
+		$prefix = 'has_tickets_';
+
+		if ( (bool) $has_tickets ) {
+			// Join to the meta that relates tickets to events but exclude RSVP tickets.
+			$repo->join_clause( "JOIN {$wpdb->postmeta} {$prefix}_ticket_event ON ( 
+					{$prefix}_ticket_event.meta_value = {$wpdb->posts}.ID
+					AND {$prefix}_ticket_event.meta_key NOT REGEXP '^_tribe_rsvp_for_event$' 
+				)" );
+
```

Comment:
:no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: a4ceda846a01f0ef1cdf2635220054cb</sub>